internal/mesh/internal/controllers/explicitdestinations/controller.go

```diff
@@ -0,0 +1,235 @@
+	destinationIDs := r.destinations.IDsForWorkload(req.ID)
+
+	decodedDestinations, err := r.fetchDestinations(ctx, rt.Client, destinationIDs)
+	if err != nil {
+		rt.Logger.Error("error fetching destinations", "error", err)
+		return err
+	}
+
+	newComputedDestinationsData := &pbmesh.ComputedExplicitDestinations{}
+	for _, dst := range decodedDestinations {
+		valid, cond := validate(ctx, rt.Client, dst)
+
+		// Only add it to computed destinations if its destinations are valid.
+		if valid {
+			newComputedDestinationsData.Destinations = append(newComputedDestinationsData.Destinations, dst.GetData().GetDestinations()...)
+		}
+
+		// Write status for this destination.
+		currentStatus := dst.GetResource().GetStatus()[ControllerName]
+		updatedStatus := &pbresource.Status{
+			Conditions:         []*pbresource.Condition{cond},
+			ObservedGeneration: dst.GetResource().GetGeneration(),
+		}
+
+		// If the status is unchanged then we should return and avoid the unnecessary write
+		if !resource.EqualStatus(currentStatus, updatedStatus, false) {
+			rt.Logger.Trace("updating status", "id", dst.GetResource().GetId())
+			_, err = rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+				Id:     dst.GetResource().GetId(),
+				Key:    ControllerName,
+				Status: updatedStatus,
+			})
+			if err != nil {
+				rt.Logger.Error("error writing new status", "id", dst.GetResource().GetId(), "error", err)
+				return err
+			}
+		}
+	}
+
+	// If after fetching and validating, we don't have any destinations,
+	// we need to skip reconcile and delete the resource.
+	if len(newComputedDestinationsData.GetDestinations()) == 0 {
+		rt.Logger.Trace("found no destinations associated with this workload")
+
+		if cpc != nil {
+			rt.Logger.Trace("deleting ComputedDestinations")
+			_, err = rt.Client.Delete(ctx, &pbresource.DeleteRequest{Id: req.ID})
+			if err != nil {
+				// If there's an error deleting CD, we want to re-trigger reconcile again.
+				rt.Logger.Error("error deleting ComputedDestinations", "error", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Lastly, write the resource.
+	if cpc == nil || !proto.Equal(cpc.GetData(), newComputedDestinationsData) {
+		rt.Logger.Trace("writing new ComputedDestinations")
+
+		// First encode the endpoints data as an Any type.
+		cpcDataAsAny, err := anypb.New(newComputedDestinationsData)
+		if err != nil {
+			rt.Logger.Error("error marshalling latest ComputedDestinations", "error", err)
+			return err
+		}
+
+		_, err = rt.Client.Write(ctx, &pbresource.WriteRequest{
+			Resource: &pbresource.Resource{
+				Id:    req.ID,
+				Owner: workloadID,
+				Data:  cpcDataAsAny,
+			},
+		})
+		if err != nil {
+			rt.Logger.Error("error writing ComputedDestinations", "error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validate(
+	ctx context.Context,
+	client pbresource.ResourceServiceClient,
+	destinations *types.DecodedDestinations) (bool, *pbresource.Condition) {
```

Comment:
Should this return signature be `(bool, []*pbresource.Condition)` instead and capture all of the reasons why it's broken instead of just the first?

Reply:
We could but the reason I changed it to a single one as opposed to how it used to be in the sidecar proxy controller is that the status type is supposed to be treated like a key in a map: https://github.com/kubernetes/community/blob/3ff3b1299e8d82993ff40dca8c4a4c9a83b5c430/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties. If we return multiple, then you could end up with multiple `DestinationsAccepted` conditions that are false. Having dynamic condition types per service didn't make sense either.

I suppose we don't have to follow that convention if we feel strongly that we want to return multiple.